### PR TITLE
[Model] Add Xing4_0 support

### PR DESCRIPTION
**Pull request by zyp2014**
## Summary

This PR adds native vLLM support for the upcoming **Xing4_0** MoE causal model.

Xing4_0 reuses the DeepSeek-V2/V3 backbone (MLA attention, MoE block, optional DSA indexer) and replaces the standard residual connection with **Manifold-constrained Hyper-Connections (mHC)**: the residual stream is expanded into `num_residual_streams` parallel streams mixed by input-dependent, doubly-stochastic matrices produced via Sinkhorn-Knopp projection. The mHC math matches the shared ops in `vllm.model_executor.layers.mhc`, so no private kernels are introduced.
Additionally, this PR enables optional FlagOS/FlagGems acceleration for Xing4_0: it enables the hot-path kernels (MoE, attention, softmax, topk, etc.) and brings measurable end-to-end inference speedups — up to 19.87% TTFT and 26.32% TPOT reduction on high-concurrency long-prompt workloads (H100 benchmark), opt-in and disabled by default.

> **Status: Draft.** Model weights are not yet public on Hugging Face Hub. This PR is opened for early code review. Once weights are released, I will:
> 1. add a test entry in `tests/models/registry.py`
> 2. update `docs/models/supported_models.md`
> 3. mark the PR ready for review.

## Changes

1. `vllm/model_executor/models/xing4_0.py`: full model implementation (`Xing4_0ForCausalLM`), including forward pass, mHC adapter, and tensor-parallel `load_weights()`.
2. Register `Xing4_0ForCausalLM` in `vllm/model_executor/models/registry.py` and related `__init__.py`.
3. Reasoning parser (`vllm/reasoning/xing4_0_reasoning_parser.py`) for reasoning-capable variants.
4. Tool parser (`vllm/tool_parsers/xing4_0_tool_parser.py`) for auto tool calling.
5. Register `xing4_0` in `vllm/config/speculative.py`, `vllm/transformers_utils/model_arch_config_convertor.py`, and `vllm/transformers_utils/config.py`.

## Optional FlagGems acceleration (opt-in)

An optional acceleration path is gated behind the env var `USE_FLAGOS=1` (disabled by default). When set, it enables a curated subset of FlagGems kernels via `flag_gems.only_enable(include=[...])`.

> **Scope note:** The acceleration is only enabled from the Xing4_0 model path — the flag_gems.only_enable(...) call lives exclusively inside xing4_0.py and is gated behind the env var, so no other model load path triggers it. The whitelist is intentionally narrow (...), and with the env var unset the behavior is fully unchanged. 
 For completeness: only_enable substitutes operators at the PyTorch dispatch level, so once enabled the kernels apply to any op in the same process that matches the whitelist — we therefore consider this an acceleration prototype and welcome feedback on a more scoped kernel backend.

```bash
pip install flagtree==0.6.1 flag-gems==5.3.4 -i https://resource.flagos.net/repository/flagos-pypi-nvidia/simple            # see version constraints in the PR discussion
USE_FLAGOS=1 vllm serve ${model_path} \
    --chat-template ${model_path}/chat_template.jinja \
    --host 0.0.0.0 \
    --port 8000 \
    --served-model-name $mode_name \
    --trust-remote-code \
    --dtype bfloat16 \
    --nnodes 1 \
    --tensor-parallel-size 1 \
    --max-model-len 262144 \
    --gpu-memory-utilization 0.90 \
    --max-num-seqs 64 \
    --max-num-batched-tokens 2048 \
    --enable-prefix-caching \
    --reasoning-parser xing4_0 \
    --enable-auto-tool-choice \
    --tool-call-parser xing4_0 \
    --uvicorn-log-level info \
    --enable-log-requests \
    --disable-custom-all-reduce
```

All Xing4_0 hot-path tests below run with FlagGems **disabled**.

## Benchmark method (H100 × 1)

- Framework: this branch vs. vanilla vLLM baseline (same branch, env var unset).
- Metric: **TTFT** (time to first token) and **TPOT** (time per output token).
- Workload: cold cache for each (input tokens + output tokens) pair.
- Each run: 5 repetitions, report the **median**; standard deviation in parentheses when ≥ 1%.
- No kernel warmup excluded; results reflect end-to-end served latency.

| input+output | concurrency | vanilla TTFT (ms) | flaggems TTFT (ms) | TTFT Δ | vanilla TPOT (ms) | flaggems TPOT (ms) | TPOT Δ |
|---|---|---|---|---|---|---|---|
| 50K+500 | 1 | 3965 | 3971 | ~0% | 10 | 10 | ~0% |
| 50K+500 | 10 | 23827 | 23050 | 3.26% | 60 | 62 | ~0% |
| 10K+500 | 1 | 486 | 483 | 0.62% | 8 | 8 | ~0% |
| 10K+500 | 10 | 1439 | 1153 | 19.87% | 38 | 28 | 26.32% |
| 1K+500 | 1 | 98 | 98 | ~0% | 8 | 7 | 12.50% |
| 1K+500 | 10 | 185 | 189 | ~0% | 17 | 17 | ~0% |

**Reading:** gains concentrate in the high-concurrency long-prompt case (10K+500, concurrency=10): ~19.87% TTFT / ~26.32% TPOT. Low-concurrency and short-prompt cases are neutral. The benefit is workload-specific, not universal.

## Test plan (internal non-public checkpoint)

1. Single-GPU: init, weight loading, text generation.
2. Tensor-parallel size=2: weight loading and inference.
3. Generation outputs and logits aligned with the Hugging Face transformers reference implementation (layer-wise comparison, reported in the accompanying notes).
4. FlagGems dual-path: default path (env unset) and accelerated path (env set) both produce consistent outputs/logits.

## Usage
```bash
vllm serve ${model_path} \
    --chat-template ${model_path}/chat_template.jinja \
    --host 0.0.0.0 \
    --port 8000 \
    --served-model-name $mode_name \
    --trust-remote-code \
    --dtype bfloat16 \
    --nnodes 1 \
    --tensor-parallel-size 1 \
    --max-model-len 262144 \
    --gpu-memory-utilization 0.90 \
    --max-num-seqs 64 \
    --max-num-batched-tokens 2048 \
    --enable-prefix-caching \
    --reasoning-parser xing4_0 \
    --enable-auto-tool-choice \
    --tool-call-parser xing4_0 \
    --uvicorn-log-level info \
    --enable-log-requests \
    --disable-custom-all-reduce
```

## Known limitations / open items for review

- **mHC checkpoint biases** (`bias_pre` / `bias_post` / `bias_res`) and the `h_res` clamp: currently merged/omitted to match the shared `MHCPreOp` signature. **Please review whether the folded formula is equivalent to the official checkpoint** — this is the main correctness item I need confirmed.
- `_xing4_0_transpose_contiguous` custom op exists to keep `comb_mix` C-contiguous for the TileLang `mhc_post` kernel that torch.compile cannot prove. Happy to simplify if reviewers prefer.
- Pipeline parallelism is not supported in mHC mode (n_streams > 1); tensor parallelism is.

## Checklist

- [x] Model registers and loads (single + TP=2)
- [x] Logit alignment vs HF reference (documented)
- [ ] Test entry in `tests/models/registry.py` (pending weights release)
- [ ] Docs entry in `docs/models/supported_models.md` (pending weights release)
- [ ] CI green

### CODE REVIEW

**Review by zyp2014 — COMMENTED** [review]
**Ready for review!**

This PR adds support for the **Xing4_0** model to vLLM.

**What's included:**

- New model implementation and weight loading under `vllm/model_executor/models/`
- Model registration / config wiring so the model is discoverable and loadable
- Docs updates: `docs/models/supported_models.md` and `docs/features/tool_calling.md`

**Testing done:**

- Verified model loads and runs forward pass / generation with the official checkpoint
- Confirmed tool calling / feature parity documented in the model card

Could you please take a look when you have time? Happy to address any feedback — especially around the weight mapping, config defaults, and CI compatibility. Thanks! 🙏
